### PR TITLE
[Merged by Bors] - chore: cleanup in Mathlib/Init/*

### PR DESCRIPTION
**Pull request by kim-em**
This moves some content out of `Init`, deprecates some material that was ported from Lean 3 but is used nowhere, and removes some attributes and instances which are duplicated from Lean or Batteries.

### CODE REVIEW

File: Mathlib/Data/List/Basic.lean [hunk 1]
```diff
@@ -33,6 +32,12 @@ universe u v w
 
 variable {ι : Type*} {α : Type u} {β : Type v} {γ : Type w} {l₁ l₂ : List α}
 
+/-- `≤` implies not `>` for lists. -/
+@[deprecated (since := "2024-07-27")]
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/List/Basic.lean:36]: Should this say what to use instead?

File: Mathlib/Init/Data/Nat/Lemmas.lean [hunk 2]
```diff
@@ -42,36 +39,26 @@ protected def ltByCases {a b : ℕ} {C : Sort u} (h₁ : a < b → C) (h₂ : a
...
-end bit
-
 /-! successor and predecessor -/
 
 def discriminate {B : Sort u} {n : ℕ} (H1 : n = 0 → B) (H2 : ∀ m, n = succ m → B) : B := by
   induction n with
   | zero => exact H1 rfl
   | succ n _ => apply H2 _ rfl
 
+@[deprecated (since := "2024-07-27")]
```
Comment (Ruben-VandeVelde) [inline Mathlib/Init/Data/Nat/Lemmas.lean:49]: Also in this file

**Review by Ruben-VandeVelde — APPROVED** [review]

File: Mathlib/Data/List/Basic.lean (hunk 1 above)
Comment (semorrison) [reply Mathlib/Data/List/Basic.lean:36]: It's just `rfl`, and it seems no one uses it, and it dates from before the port. 🤷‍♀️

File: Mathlib/Logic/Basic.lean [hunk 3]
```diff
@@ -767,6 +770,32 @@ def choice_of_byContradiction' {α : Sort*} (contra : ¬(α → False) → α) :
 lemma choose_eq' (a : α) : @Exists.choose _ (a = ·) ⟨a, rfl⟩ = a :=
   (@choose_spec _ (a = ·) _).symm
 
+alias axiom_of_choice := axiomOfChoice -- TODO: remove? rename in core?
+alias by_cases := byCases -- TODO: remove? rename in core?
+alias by_contradiction := byContradiction -- TODO: remove? rename in core?
+
+-- The remaining theorems in this section was ported from Lean 3,
```
Comment (grunweg) [inline Mathlib/Logic/Basic.lean:777]:
```suggestion
-- The remaining theorems in this section were ported from Lean 3,

File: Mathlib/Init/Data/Fin/Basic.lean [hunk 4]
```diff
@@ -26,9 +26,9 @@ theorem eq_of_veq : i.val = j.val → i = j := eq_of_val_eq
 @[deprecated val_eq_of_eq (since := "2024-02-15")]
 theorem veq_of_eq : i = j → i.val = j.val := val_eq_of_eq
 
--- These two aren't deprecated because `ne_of_val_ne` and `val_ne_of_ne`
--- use `¬a = b` instead of `a ≠ b`. TODO: fix or rename in Lean core.
+@[deprecated ne_of_val_ne (since := "2024-07-27")]
```
Comment (grunweg) [inline Mathlib/Init/Data/Fin/Basic.lean:29]: Nice clean-up.

File: Mathlib/Order/RelClasses.lean [hunk 5]
```diff
@@ -244,6 +244,30 @@ lemma WellFounded.prod_lex {ra : α → α → Prop} {rb : β → β → Prop} (
...
+
+open PSigma
+
+/-- The lexicographical order of well-founded relations is well-founded. -/
+theorem WellFounded.psigma_lex
+    {α : Sort*} {β : α → Sort*} {r : α → α → Prop} {s : ∀ a : α, β a → β a → Prop}
+    (ha : WellFounded r) (hb : ∀ x, WellFounded (s x)) : WellFounded (Lex r s) :=
+  WellFounded.intro fun ⟨a, b⟩ => lexAccessible (WellFounded.apply ha a) hb b
+
+theorem WellFounded.psigma_revLex
```
Comment (grunweg) [inline Mathlib/Order/RelClasses.lean:257]: This also was phrased for Sort*, can you re-add this (or say why this is not desirable), please?

File: Mathlib/Init/Logic.lean [hunk 6]
```diff
@@ -78,15 +76,8 @@ attribute [refl] Iff.refl
 attribute [trans] Iff.trans
 attribute [symm] Iff.symm
 
--- This is needed for `calc` to work with `iff`.
-instance : Trans Iff Iff Iff where
```
Comment (grunweg) [inline Mathlib/Init/Logic.lean:82]: Just flagging this for other reviewers: you are deleting this instance. Do we have a test this works?

File: Mathlib/Init/Data/Sigma/Lex.lean [hunk 7]
```diff
@@ -1,61 +0,0 @@
...
-
-section
-
-theorem skipLeft_wf (α : Type u) {β : Type v} {s : β → β → Prop} (hb : WellFounded s) :
-    WellFounded (SkipLeft α s) :=
-  revLex_wf emptyWf.wf hb
-
-end
-
-instance hasWellFounded {α : Type u} {β : α → Type v} [s₁ : WellFoundedRelation α]
```
Comment (grunweg) [inline Mathlib/Init/Data/Sigma/Lex.lean:56]: What happened to this instance?

File: Mathlib/Init/Data/Nat/Lemmas.lean [hunk 8]
```diff
@@ -42,36 +39,26 @@ protected def ltByCases {a b : ℕ} {C : Sort u} (h₁ : a < b → C) (h₂ : a
...
 /-! induction principles -/
 
-
 def twoStepInduction {P : ℕ → Sort u} (H1 : P 0) (H2 : P 1)
     (H3 : ∀ (n : ℕ) (_IH1 : P n) (_IH2 : P (succ n)), P (succ (succ n))) : ∀ a : ℕ, P a
   | 0 => H1
   | 1 => H2
   | succ (succ _n) => H3 _ (twoStepInduction H1 H2 H3 _) (twoStepInduction H1 H2 H3 _)
 
+@[deprecated (since := "2024-07-27")]
```
Comment (grunweg) [inline Mathlib/Init/Data/Nat/Lemmas.lean:61]: I presume this is also just unused: can you add a comment as in the other file?

File: Mathlib/Init/Data/Nat/Lemmas.lean [hunk 9]
```diff
@@ -93,16 +80,10 @@ protected theorem case_strong_induction_on {p : Nat → Prop} (a : Nat) (hz : p
 
 /-! mod -/
 
+@[deprecated (since := "2024-07-27")]
 theorem cond_decide_mod_two (x : ℕ) [d : Decidable (x % 2 = 1)] :
```
Comment (grunweg) [inline Mathlib/Init/Data/Nat/Lemmas.lean:84]: Same here.

**Review by grunweg — COMMENTED** [review]
I looked over everything. Thanks a lot for doing this clean-up.
I trust you that all attributes you're removing are in fact present in batteries or core.
Apart from that, I only have a few minor comments.

File: Mathlib/Init/Logic.lean (hunk 6 above)
Comment (semorrison) [reply Mathlib/Init/Logic.lean:81]: Yes, it is used in a few places, e.g. Mathlib.Topology.Order.LeftRightNhds and Archive.Imo.Imo1964Q1

**Comment by grunweg** [comment]
Thanks for addressing my comments (and for performing this clean-up in the first place). Let's have another pair of eyes on this.
maintainer merge

File: Mathlib/Init/Data/Sigma/Lex.lean (hunk 7 above)
Comment (semorrison) [reply Mathlib/Init/Data/Sigma/Lex.lean:56]: This was a duplicate of `PSigma.instWellFoundedRelation` in `lean4`.

Comment (grunweg) [reply Mathlib/Init/Data/Sigma/Lex.lean:56]: :+1: ... and the `lean4` instance even applies to Sort. Thanks for explaining!

**Review by sgouezel — COMMENTED** [review]
bors d+
Thanks!

**Comment by YaelDillies** [comment]
This is clearly not a `feat` PR, right?

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**